circuit-lib/circuit-lib.js/src/elgamal-babyjubjub/elgamal.ts

```diff
@@ -0,0 +1,287 @@
+  const ephemeral_key = babyJub.BASE.multiply(nonce);
+  const masking_key = pubKey.multiply(nonce);
+  let encrypted_message: BabyJubExtPoint;
+  // The sender encrypts the encodedMessage
+  if (pubKey.assertValidity! && !pubKey.equals(babyJub.ZERO)) {
+    encrypted_message = message.add(masking_key);
+  } else throw new Error("Invalid Public Key!");
+
+  return { ephemeral_key, encrypted_message };
+}
+
+/**
+ * Decrypts a ciphertext using a private key.
+ * @param privKey The private key
+ * @param ciphertext The ciphertext to decrypt
+ */
+function decrypt(
+  privKey: PrivKey,
+  ephemeral_key: BabyJubExtPoint,
+  encrypted_message: BabyJubExtPoint,
+): BabyJubExtPoint {
+  // The receiver decrypts the message => encryptedMessage - [privKey].ephemeralKey
+  const masking_key = ephemeral_key.multiply(
+    ElGamalUtils.formatPrivKeyForBabyJub(privKey),
+  );
+  const decrypted_message = encrypted_message.add(masking_key.negate());
+
+  return decrypted_message;
+}
+
+/**
+ * Randomize a ciphertext such that it is different from the original
+ * ciphertext but can be decrypted by the same private key.
+ * @param pubKey The same public key used to encrypt the original encodedMessage
+ * @param ciphertext The ciphertext to re-randomize.
+ * @param randomVal A random value z such that the re-randomized ciphertext could have been generated a random value y+z in the first
+ *                  place (optional)
+ */
+function rerandomize(
+  pubKey: PubKey,
+  ephemeral_key: BabyJubExtPoint,
+  encrypted_message: BabyJubExtPoint,
+  randomVal?: bigint,
+) {
+  const nonce = randomVal ?? genRandomSalt();
+  const randomized_ephemeralKey = ephemeral_key.add(
+    babyJub.BASE.multiply(nonce),
+  );
+
+  const randomized_encryptedMessage = encrypted_message.add(
+    pubKey.multiply(nonce),
+  );
+
+  return { randomized_ephemeralKey, randomized_encryptedMessage };
+}
+
+namespace ElGamalUtils {
+  export const stringifyBigInts: (obj: object) => any =
+    ff.utils.stringifyBigInts;
+  export const unstringifyBigInts: (obj: object) => any =
+    ff.utils.unstringifyBigInts;
+
+  /**
+   * A utility function for public key derivation.
+   * Convert a BigInt to a Buffer
+   */
+  export const bigInt2Buffer = (i: BigInt): Buffer => {
+    return Buffer.from(i.toString(16), "hex");
+  };
+
+  /**
+   * A utility function for public key derivation.
+   * Taken from https://github.com/iden3/circomlibjs/blob/main/src/eddsa.js
+   */
+  export function prv2pub(prv: any) {
+    const sBuff = pruneBuffer(
+      createBlakeHash("blake512").update(Buffer.from(prv)).digest(),
+    );
+    let s = Scalar.fromRprLE(sBuff, 0, 32);
+    const A = babyJub.BASE.multiply(BigInt(Scalar.shr(s, 3)));
+    return A;
+  }
+
+  /**
+   * A utility function for formatting a Baby Jubjub private key.
+   * Taken from https://github.com/iden3/circomlibjs/blob/main/src/eddsa.js
+   */
+  export function pruneBuffer(buff: any) {
```

Comment:
Do you understand what this function does?

Reply:
Removed after replacing code with imported functions from [maci-crypto](https://www.npmjs.com/package/maci-crypto)
The function is used to remove the excess bytes after converting a `BigInt` into a `Buffer`.